### PR TITLE
Add bookmarks

### PR DESCRIPTION
**Pull request by ClearlyClaire**
Add a bookmark feature, working mostly like favourites, except they are not federated nor visible by other users than the person bookmarking a toot.

### CODE REVIEW

**Comment by m4sk1n** [comment]
I think the “favourite” thing functionality should be reduced now to work like a zucksite like button, without the save feature, but then it would be nice to have an activity log, so there would still be way to find favourited toots

**Comment by Cassolotl** [comment]
This seems like a good change, I like private save-for-later as separate from public reaction hearts/stars.

Related issues:

* Dream feature: "like" and "save" separately #1178
* arbitrary emojis as reactions #4106 - maybe the heart/star can just be like a positive reaction button and it doesn't need to be saved?

**Comment by trwnh** [comment]
re: what m4sk1n said -- i don't see a compelling reason to change anything about how faves work currently. fundamentally, they are a public approval / acknowledgement of a message, and can be browsed from someone's profile if they want to see go back and see them.

a bookmark feature would be almost the same thing, but private -- no indicator, not exposed via the profile or the status. adding bookmarks is no reason to remove functionality from the favorites.

**Comment by Cassolotl** [comment]
I feel like "like/star" doesn't really need to be saved in a list if it's just a reaction emoji ish thing - if folks want to save it for later they'll use the bookmark feature, right?

And I think probably bookmarked toots should show up in search because it's save-for-later and that'll be something you want to be able to look up later.

I think "like/star" toots don't need to show up in a search because it's just public approval and that could clog up the search results unnecessarily - most people are gonna "like/star" *much* more than they'll save-for-later, and they're less likely to need to pull liked toots up in searches later, right?

**Comment by nightpool** [comment]
disagree with adding this per my comments on #1178.

**Comment by nightpool** [comment]
Hmm. Not sure what happened to that issue, it seems to be completely deleted, with it's discussion.

**Comment by ThibG** [comment]
Ok, your main argument seems to be “web browser bookmarks serve this purpose”, yes and no.
If you bookmark the public toot, it's a pain to get back into the WebUI to boost/like/reply to it.
If you bookmark the WebUI, it's a bit better, but you go through the trouble of reloading the whole WebUI, either in an extra tab, or replacing your position in timelines, etc.

**Comment by ThibG** [comment]
@lanodan favorites are already searchable (along with the rest) if Full-Text-Search is enabled.
The point of the “bookmark” feature as I use it is to bookmark things I haven't the time to read right now but I potentially want to react on later. For instance, if someone posts a long article or a video, and I want to like/boost/comment once I've seen it.

**Comment by Gargron** [comment]
@ThibG But I use favourites for that use case. And that having a social function makes sense to me, I remember the old Delicious used to display rankings of favourited links and stuff that authors of those links could be proud of. I'm not a big fan of splitting this functionality. In fact, I was not a fan when Twitter changed favourites to likes, and now iirc *they* are adding "bookmarks" and I think that's a hot mess that could've been avoided by just keeping favourites the way they are here right now.

**Comment by ThibG** [comment]
@Gargron but why should that person be proud of something I haven't read yet? Afaik, people use favorites (or likes, that's what they are protocol-wise) as approval, encouragement, or “I read this” marks. A bookmark is “I want to read this” and is something completely different honestly.

**Comment by ThibG** [comment]
Worth a shot I guess :man_shrugging:

pr

**Comment by ThibG** [comment]
Updated the PR to:
- Rebase it
- Remove the joke rainbows
- Add OAuth scopes for bookmarks

**Comment by htgoebel** [comment]
I really like to see this featur merged! I'm bookmarking a lot in Fedilab which I want to read later on the desktop (using the WebUI).

**Comment by ThibG** [comment]
Twitter has them now :3

**Comment by regagain** [comment]
I'd still like to have that feature.

**Comment by umonaca** [comment]
Tried on local dev environment with some existing data, but the database migration failed.
`rails db:migrate --trace` shows:

```
rails db:migrate --trace
** Invoke db:migrate (first_time)
** Invoke db:load_config (first_time)
** Invoke environment (first_time)
** Execute environment
** Execute db:load_config
** Execute db:migrate
   (0.2ms)  SELECT pg_try_advisory_lock(5881450594042647960)
   (0.7ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
Migrating to CreateBookmarks (20180831171112)
   (0.2ms)  BEGIN
Chewy strategies stack: [2] <- bypass @ /home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/railtie.rb:37
== 20180831171112 CreateBookmarks: migrating ==================================
-- create_table(:bookmarks, {})
   (60.1ms)  CREATE TABLE "bookmarks" ("id" bigserial primary key, "account_id" bigint NOT NULL, "status_id" bigint NOT NULL, "created_at" timestamp NOT NULL, "updated_at" timestamp NOT NULL)
   (29.1ms)  CREATE  INDEX  "index_bookmarks_on_account_id" ON "bookmarks"  ("account_id")
   (30.0ms)  CREATE  INDEX  "index_bookmarks_on_status_id" ON "bookmarks"  ("status_id")
   -> 0.1280s
Chewy strategies stack: [2] -> bypass, now bypass @ /home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/railtie.rb:37
   (1.5ms)  ROLLBACK
   (0.6ms)  SELECT pg_advisory_unlock(5881450594042647960)
rails aborted!
StandardError: An error has occurred, this and all later migrations canceled:

=== Dangerous operation detected #strong_migrations ===

New foreign keys are validated by default. This acquires an AccessExclusiveLock,
which is expensive on large tables. Instead, validate it in a separate migration
with a more agreeable RowShareLock.

class StrongMigrations::Checker < ActiveRecord::Migration[5.2]
  def change
    add_foreign_key :bookmarks, :accounts, column: :account_id, on_delete: :cascade, validate: false
  end
end

class ValidateStrongMigrations::Checker < ActiveRecord::Migration[5.2]
  def change
    validate_foreign_key :bookmarks, :accounts
  end
end

/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/strong_migrations-0.4.2/lib/strong_migrations/migration.rb:21:in `stop!'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/strong_migrations-0.4.2/lib/strong_migrations/checker.rb:229:in `raise_error'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/strong_migrations-0.4.2/lib/strong_migrations/checker.rb:148:in `perform'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/strong_migrations-0.4.2/lib/strong_migrations/migration.rb:9:in `method_missing'
/home/mastodon/live/db/migrate/20180831171112_create_bookmarks.rb:10:in `change'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:814:in `exec_migration'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:798:in `block (2 levels) in migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/2.6.0/benchmark.rb:293:in `measure'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:797:in `block in migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/connection_adapters/abstract/connection_pool.rb:416:in `with_connection'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:796:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/strong_migrations-0.4.2/lib/strong_migrations/migration.rb:5:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/railtie.rb:37:in `block in migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/strategy.rb:70:in `wrap'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy.rb:201:in `strategy'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/railtie.rb:37:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:977:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1292:in `block in execute_migration_in_transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1343:in `block in ddl_transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `block in transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/connection_adapters/abstract/transaction.rb:239:in `block in within_new_transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/connection_adapters/abstract/transaction.rb:236:in `within_new_transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/connection_adapters/abstract/database_statements.rb:267:in `transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/transactions.rb:212:in `transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1343:in `ddl_transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1291:in `execute_migration_in_transaction'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1263:in `block in migrate_without_lock'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1262:in `each'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1262:in `migrate_without_lock'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1210:in `block in migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1363:in `with_advisory_lock'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1210:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/railtie.rb:37:in `block in migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/strategy.rb:70:in `wrap'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy.rb:201:in `strategy'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/chewy-5.1.0/lib/chewy/railtie.rb:37:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1036:in `up'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/migration.rb:1011:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/tasks/database_tasks.rb:172:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/strong_migrations-0.4.2/lib/strong_migrations/database_tasks.rb:4:in `migrate'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activerecord-5.2.3/lib/active_record/railties/databases.rake:60:in `block (2 levels) in <top (required)>'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `block in execute'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `each'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/task.rb:273:in `execute'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/task.rb:214:in `block in invoke_with_call_chain'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/2.6.0/monitor.rb:235:in `mon_synchronize'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/task.rb:194:in `invoke_with_call_chain'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/task.rb:183:in `invoke'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/application.rb:160:in `invoke_task'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `block (2 levels) in top_level'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `each'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/application.rb:116:in `block in top_level'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/application.rb:125:in `run_with_threads'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/application.rb:110:in `top_level'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-5.2.3/lib/rails/commands/rake/rake_command.rb:23:in `block in perform'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/rake-12.3.3/lib/rake/application.rb:186:in `standard_exception_handling'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-5.2.3/lib/rails/commands/rake/rake_command.rb:20:in `perform'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-5.2.3/lib/rails/command.rb:48:in `invoke'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-5.2.3/lib/rails/commands.rb:18:in `<top (required)>'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `block in require_with_bootsnap_lfi'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:21:in `require_with_bootsnap_lfi'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.5/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:30:in `require'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `block in require'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:257:in `load_dependency'
/home/mastodon/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-5.2.3/lib/active_support/dependencies.rb:291:in `require'
bin/rails:4:in `<main>'
Tasks: TOP => db:migrate
```

**Comment by umonaca** [comment]
Update: it's the strong migration issue.
To anyone who may be interested in this PR: you may wrap the `add_foreign_key` lines with `safety_assured` or follow [the guide here](https://github.com/ankane/strong_migrations) to create a separate migration.

**Comment by Gargron** [comment]
Okay, I've heard some good arguments for having this feature today. This PR needs to be cleaned up a bit to be merged.

**Comment by ThibG** [comment]
Rebased and fixed the migration. Would have to go other it to make sure everything has been kept up-to-date.

File: app/javascript/mastodon/features/getting_started/index.js [hunk 1]
```diff
@@ -22,6 +22,7 @@ const messages = defineMessages({
   settings_subheading: { id: 'column_subheading.settings', defaultMessage: 'Settings' },
   community_timeline: { id: 'navigation_bar.community_timeline', defaultMessage: 'Local timeline' },
   direct: { id: 'navigation_bar.direct', defaultMessage: 'Direct messages' },
+  bookmarks: { id: 'navigation_bar.bookmarks', defaultMessage: 'Bookmarks' },
```
Comment (umonaca) [inline app/javascript/mastodon/features/getting_started/index.js]: The navigation bar item is missing. You should probably add something like this to `/home/muveexu/donotban/live/app/javascript/mastodon/features/ui/components/navigation_panel.js`
```
<NavLink className='column-link column-link--transparent' to='/bookmarks'><Icon className='column-link__icon' id='bookmark' fixedWidth /><FormattedMessage id='navigation_bar.bookmarks' defaultMessage='Bookmarks' /></NavLink>    
```
And adjust the column height for single-column mode accordingly:
Change [this line](https://github.com/tootsuite/mastodon/blob/c31bc41a02d89eea654f8d7c74c0a1cb544cd919/app/javascript/mastodon/features/getting_started/index.js#L133) to `height += 48*4;`.

**Review by umonaca — CHANGES_REQUESTED** [review]

**Comment by umonaca** [comment]
You would probably want [this commit](https://github.com/glitch-soc/mastodon/pull/1166) as well

**Comment by umonaca** [comment]
Bug report: the un-bookmark is not as responsive as the un-favourite. When a toot is unbookmarked, the icon remains `active` so the user is unsure if the toot has been unbookmarked or not. The user has to refresh the entire page to see the unbookmark has succeeded.

I reported [this issue}(https://github.com/glitch-soc/mastodon/issues/1244) in glitch-soc but was unable to reproduce it on the glitch-soc instance. However, I am able to reliably reproduce this bug in this PR in a local test environment.

File: app/javascript/mastodon/actions/interactions.js [hunk 2]
```diff
@@ -187,6 +195,76 @@ export function unfavouriteFail(status, error) {
...
+    });
+  };
+};
+
+export function unbookmark(status) {
+  return (dispatch, getState) => {
+    dispatch(unbookmarkRequest(status));
+
+    api(getState).post(`/api/v1/statuses/${status.get('id')}/unbookmark`).then(response => {
+      dispatch(unbookmarkSuccess(status, response.data));
```
Comment (umonaca) [inline app/javascript/mastodon/actions/interactions.js]: Update: simply adding `dispatch(importFetchedStatus(response.data));` to this line and the `bookmark(status)` function  fixed the bug I reported.

Comment (umonaca) [inline app/javascript/mastodon/actions/interactions.js]: Update: simply adding `dispatch(importFetchedStatus(response.data));` to this line and the `bookmark(status)` function  fixed the bug I reported.

**Review by umonaca — CHANGES_REQUESTED** [review]

File: app/javascript/mastodon/components/status_action_bar.js [hunk 3]
```diff
@@ -114,6 +116,14 @@ class StatusActionBar extends ImmutablePureComponent {
     window.open(`/interact/${this.props.status.get('id')}?type=${type}`, 'mastodon-intent', 'width=445,height=600,resizable=no,menubar=no,status=no,scrollbars=yes');
   }
 
+  handleBookmarkClick = () => {
+    this.props.onBookmark(this.props.status);
+  }
+
+  handleReblogClick = (e) => {
```
Comment (umonaca) [inline app/javascript/mastodon/components/status_action_bar.js]: Is there any reason to have two `handleReblogClick` in this file?

File: app/javascript/mastodon/features/status/components/action_bar.js [hunk 4]
```diff
@@ -198,6 +204,7 @@ class ActionBar extends React.PureComponent {
         <div className='detailed-status__button'><IconButton disabled={reblog_disabled} active={status.get('reblogged')} title={reblog_disabled ? intl.formatMessage(messages.cannot_reblog) : intl.formatMessage(messages.reblog)} icon={reblogIcon} onClick={this.handleReblogClick} /></div>
         <div className='detailed-status__button'><IconButton className='star-icon' animate active={status.get('favourited')} title={intl.formatMessage(messages.favourite)} icon='star' onClick={this.handleFavouriteClick} /></div>
         {shareButton}
+        <div className='detailed-status__button'><IconButton className='bookmark-icon' active={status.get('bookmarked')} title={intl.formatMessage(messages.bookmark)} icon='bookmark' onClick={this.handleBookmarkClick} /></div>
```
Comment (Gargron) [inline app/javascript/mastodon/features/status/components/action_bar.js]: I heard at least one user opinion that favourites and bookmarks not need be side-by-side, bookmarks can be in the dropdown menu. More responses have stated that they would use bookmarks more rarely than favourites, so perhaps this makes sense? Thoughts?

Comment (Gargron) [reply app/javascript/mastodon/features/status/components/action_bar.js]: Oh well, we can move it later.

**Review by Gargron — APPROVED** [review]

**Comment by bannisdale** [comment]
Maybe a stupid question, but ... is this already live? Is this part of the next release? How would I be able to find out? :)

**Comment by Gargron** [comment]
It's live on mastodon.social, but no stable version has been released since this PR was merged, yet.

**Pull request merged**